keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,388 @@
+  operator instance
+- MUST include application-related diagnostics artifacts in the archive
+- MUST include instance-related diagnostics artifacts in the archive
+- MUST include KUDO-related diagnostics artifacts in the archive
+- MUST include "Kubernetes workload"-related diagnostics artifacts in the
+  archive
+- MUST accept parameters and work without interactive prompts
+- SHOULD work in airgapped environments
+- SHOULD report the versions of every component and tool, in the archive (e.g.,
+  the version of the collector, the application version, the operator version,
+  the KUDO version, the Kubernetes version, etc.)
+- SHOULD follow Kubernetes' ecosystem conventions and best practices
+- MUST be published as a static binary
+- SHOULD make it possible to publish archive to cloud object storage (AWS S3,
+  etc.)
+- MUST follow SemVer
+
+## Proposal
+
+### Operator user experience
+
+The output from diagnostics collection is an archive containing all
+diagnostics artifacts for the provided operator instance.
+
+```bash
+kubectl kudo diagnostics collect --instance=%instance% --namespace=%namespace%
+```
+
+### Operator developer experience
+
+To configure diagnostics for a given operator version, this KEP introduces an optional top-level
+`diagnostics` key in operator.yaml.
+
+#### Diagnostics collection
+
+The following diagnostics will be implicitly collected without any configuration
+from the operator developer:
+
+- Logs for deployed pods related to the KUDO Instance
+- YAML for created resources, including both spec and status, related
+  to the KUDO instance
+- Output of `kubectl describe` for all deployed resources related to the KUDO
+  Instance
+- Current plan status, if one exists, or the KUDO Instance
+- Information about the KUDO instance's Operator and OperatorVersion
+- Logs for the KUDO controller manager
+- Describe for the KUDO controller manager resources
+- RBAC resources that are applicable to the KUDO controller manager
+- Current settings and version information for KUDO
+- k8s events (can we filter them for resources that the instance owns?)
+
+Operator developer experience, then, focuses on customizing diagnostics
+information to gather information about the running application. The following
+forms are available, subject to change over time:
+
+- **Copy**: Copy a file out of a running pod. This is useful for non-stdout/err
+  logs, configuration files, and other artifacts generated by an application.
+  Higher level resources can also be used, which will copy the file on all pods
+  selected by that resource.
+- **Command**: Run a command on a running pod and copy the stdout/err. Higher level
+  resources can also be used, which will run the command on all pods selected by
+  that resource.
+- **Task**: Run a KUDO task and copy the stdout and other arbitrary files.
+- **HTTP**: Make an HTTP request from the KUDO controller manager to a named
```

Comment:
I'm not sure we want to promise that it will be the controller manager who does the fetching... Perhaps this should just say it should be fetched by a process running within the cluster?